srv/modules/runners/disk_led.py

```diff
@@ -0,0 +1,208 @@
+# -*- coding: utf-8 -*-
+"""
+This runner is used to turn on/off the ident and fault LED of a disk.
+"""
+from __future__ import absolute_import
+
+import logging
+import pprint
+import re
+
+import salt.client
+
+log = logging.getLogger(__name__)
+
+
+def _process(hostname, device_name, led, status):
+    """
+    Turn on/off the specified LED of the disk on the given host.
+    """
+    local_client = salt.client.LocalClient()
+    device_file = '/dev/{}'.format(device_name)
+    status = 'on' if status in [True, 'on'] else 'off'
+
+    # Get the configuration of the command to turn on/off the ident
+    # or fault LED on the target host.
+    cmd_cfg = local_client.cmd(
+        hostname, 'pillar.get', ['disk_led:cmd'],
+        tgt_type='compound')[hostname]
+
+    print("Turning {} the {} LED of disk '{}' on host '{}' ...".format(
+        status, led, device_file, hostname))
+    _cmd_run(hostname, cmd_cfg[led][status].format(device_file=device_file))
+
+
+def _cmd_run(hostname, cmd):
+    """
+    Run the specified command line.
+    .. note:: This function has been introduced for easier unit testing.
+    """
+    local_client = salt.client.LocalClient()
+    local_client.cmd(hostname, 'cmd.run', [cmd], tgt_type='compound')
+
+
+def _get_metadata():
+    """
+    Get the OSD metadata. We have to collect this data on the master
+    because it is mostly not possible to collect this data on storage
+    nodes due to missing privileges.
+    :return: Returns the OSD metadata.
+    :rtype: list
+    """
+    __master_opts__ = salt.config.client_config("/etc/salt/master")
+    __master_utils__ = salt.loader.utils(__master_opts__)
+    __salt_master__ = salt.loader.minion_mods(
+        __master_opts__, utils=__master_utils__)
+    master_minion = __salt_master__['master.minion']()
+
+    local_client = salt.client.LocalClient()
+    return local_client.cmd(master_minion, 'osd.metadata', tgt_type='compound')
+
+
+def osd(osd_id, device_type, led, status):
+    """
+    Manage the storage enclosure LED by OSD identifier.
+    :param osd_id: The OSD id, e.g.
+        - 4
+        - osd.1
+    :type osd_id: str
+
+    :param device_type: The device to use. This can be:
+        - primary (Blue-/Filestore)
+        - wal (Bluestore)
+        - db (Bluestore)
+        - journal (Filestore)
+        - all
+    :type device_type: str
+    :param led: The LED to be processed. This can be 'ident' or 'fault'.
+    :type led: str
+    :param status: Set to 'on' to switch on the light, otherwise 'off'.
+    :type status: str
+    """
+    osd_id = int(re.sub(r'^osd\.', '', str(osd_id)))
+    local_client = salt.client.LocalClient()
+
+    osd_metadata = _get_metadata()
```

Comment:
That returns in the form of `{'host': [{content}]}`.

You are passing the data straight down to `osd.details` where you do a `for osd_metadata_v in osd_metadata:`.
The passed metadata is not a `list` though.

What you are looking for is the `values` of the metadata.

``` python
list(osd_metadata.values())
```
I'd also suggest to validate all output you get from minions. There is always a chance you get bogus returns due to several reasons (modules not synced, timeouts etc). All consecutive calls would fall with the first error which makes it also very hard to debug

Reply:
Fixed